File: app/src/main/java/com/microlabs/trallet/repo/DatabaseBookRepository.java
package com.microlabs.trallet.repo;

import com.microlabs.trallet.model.Book;
import com.microlabs.trallet.model.Category;
import com.microlabs.trallet.model.Currency;
import com.microlabs.trallet.model.Expense;

import java.util.Date;
import java.util.List;

import io.realm.Realm;
import io.realm.Sort;

/**
 * Created by Hansen on 4/2/2017.
 * <p>
 * Database of Book and Expense Repository
 */

public class DatabaseBookRepository {

    private Realm realm;

    /**
     * Get Default Instance of Realm Database
     */
    public DatabaseBookRepository() {
        realm = Realm.getDefaultInstance();
    }

    /**
     * Get All Book List sorted by its Id
     *
     * @return list of Book, sort by its Id ascending
     */
    public List<Book> getBookList() {
        return realm.where(Book.class).findAllSorted(Book.fId);
    }

    /**
     * Check whether default Category is exist
     * Default Category is 4 (Food, Transport, Shop, and Others)
     *
     * @return true if default Category is NOT exists
     */
    public boolean isDefaultCategoryNotExist() {
        return realm.where(Category.class).count() < 4;
    }

    /**
     * Check whether default Currency is exist
     * Default Currency is 2 (USD $ and IDR Rp)
     *
     * @return true if Default Currency is NOT exists
     */
    public boolean isDefaultCurrencyNotExist() {
        return realm.where(Currency.class).count() < 2;
    }

    /**
     * Insert Default Category to Repository if not exists
     */
    public void loadDefaultCategory() {
        for (int i = 0; i < 4; i++) {
            realm.beginTransaction();
            Category c = realm.createObject(Category.class);
            c.setId(realm);
            switch (i) {
                case 0:
                    c.setName("Foods");
                    break;
                case 1:
                    c.setName("Transport");
                    break;
                case 2:
                    c.setName("Shop");
                    break;
                case 3:
                    c.setName("Others");
                    break;
            }
            realm.commitTransaction();
        }
    }

    /**
     * Insert Default Currency if not exists
     */
    public void loadDefaultCurrency() {
        for (int i = 0; i < 2; i++) {
            realm.beginTransaction();
            Currency c = realm.createObject(Currency.class);
            c.setId(realm);
            switch (i) {
                case 0:
                    c.setName("IDR Rp");
                    c.setValue(1.0);
                    break;
                case 1:
                    c.setName("USD $");
                    c.setValue(13500.0);
                    break;
            }
            realm.commitTransaction();
        }
    }

    /**
     * Get List of Expenses from a book
     *
     * @param bookId = bookId where List of Expenses is
     * @return = List of Expenses, sorted by its Id descending (latest is first)
     */
    public List<Expense> getExpenseList(int bookId) {
        return realm.where(Expense.class).equalTo(Expense.fBook, bookId).findAllSorted(Expense.fId, Sort.DESCENDING);
    }

    /**
     * Delete a Book by its Id
     *
     * @param bookId = bookId of Book to remove
     */
    public void deleteBook(int bookId) {
        RealmHelper.deleteBook(realm, bookId);
    }

    /**
     * Delete an Expense by its Id
     *
     * @param bookId    = bookId of Expense to remove
     *                  Used to decrease book total value
     * @param expenseId = expenseId of Expense to remove
     */
    public void deleteExpense(int bookId, int expenseId) {
//        RealmHelper.setBookDelete(realm, bookId, expenseId);
        RealmHelper.deleteExpense(realm, expenseId);
    }

    /**
     * Get Expense Object by its Id
     *
     * @param id = expenseId to get
     * @return = expense object found
     */
    public Expense getExpense(int id) {
        return realm.where(Expense.class).equalTo(Expense.fId, id).findFirst();
    }

    /**
     * Save Expense to database
     * and set its Book total value
     *
     * @param title      = Expense title
     * @param bookId     = Expense bookId
     * @param value      = Expense value
     * @param categoryId = Expense CategoryId
     * @param currencyId = Expense CurrencyId
     * @param date       = Expense Date
     * @param details    = Expense Details
     */
    public void saveExpense(String title, int bookId, Double value, int categoryId, int currencyId, Date date, String details) {
        realm.beginTransaction();
        Expense expense = realm.createObject(Expense.class);
        expense.setId(realm);
        expense.setTitle(title);
        expense.setValue(value);
        expense.setCategoryId(categoryId);
        expense.setCurrencyId(currencyId);
        expense.setDetails(details);
        expense.setBookId(bookId);
        expense.setDate(date);
        realm.commitTransaction();
//        RealmHelper.setBookTotal(realm, bookId, value);
    }

    /**
     * Close Realm Instance to prevent memory leaks
     */
    public void close() {
        realm.close();
    }

    /**
     * Get All Category
     *
     * @return = List of Category
     */
    public List<Category> getCategoryList() {
        return realm.where(Category.class).findAll();
    }

    /**
     * Get All Currency
     *
     * @return = List of Currency
     */
    public List<Currency> getCurrencyList() {
        return realm.where(Currency.class).findAll();
    }

    /**
     * Update Expense Data
     *
     * @param expenseId  = expenseId to be updated
     * @param title      = updated expense title
     * @param bookId     = expense bookId to update the book total value
     * @param value      = updated expense value
     * @param categoryId = updated expense categoryId
     * @param currencyId = updated expense currencyId
     * @param date       = updated expense date
     * @param details    = updated expense details
     * @param oldValue   = expense old value to update the book total value
     */
    public void updateExpense(int expenseId, String title, int bookId, double value, int categoryId, int currencyId, Date date, String details, double oldValue) {
        realm.beginTransaction();
        final Expense expense = realm.where(Expense.class).equalTo(Expense.fId, expenseId).findFirst();
        expense.setTitle(title);
        expense.setValue(value);
        expense.setDetails(details);
        expense.setCategoryId(categoryId);
        expense.setCurrencyId(currencyId);
        expense.setDate(date);
        realm.commitTransaction();
//        RealmHelper.setBookEdit(realm, bookId, oldValue, value);
    }

    /**
     * Save New Currency
     *
     * @param title = Currency title
     * @param value = Currency value
     * @return = true if success to save, false if currency title already exists
     */
    public boolean saveCurrency(String title, double value) {
        if (realm.where(Currency.class).equalTo(Currency.fName, title).findFirst() == null) {
            realm.beginTransaction();
            Currency currency = realm.createObject(Currency.class);
            currency.setId(realm);
            currency.setName(title);
            currency.setValue(value);
            realm.commitTransaction();
            return true;
        } else {
            return false;
        }
    }

    /**
     * Update Currency Details
     *
     * @param id    = currencyId of Currency to be updated
     * @param title = currency title
     * @param value = currency value
     * @return true if success to update currency, fail if other currency with same title already exists
     */
    public boolean updateCurrency(int id, String title, double value) {
        if (realm.where(Currency.class).equalTo(Currency.fName, title).notEqualTo(Currency.fId, id).findFirst() != null) {
            realm.beginTransaction();
            Currency currency = realm.where(Currency.class).equalTo(Currency.fId, id).findFirst();
            currency.setName(title);
            currency.setValue(value);
            realm.commitTransaction();
            return true;
        } else {
            return false;
        }
    }

    /**
     * Check if currency can be removed
     * currency can only be removed when no longer being used
     *
     * @param id = currencyId to be checked
     * @return true if currency can be removed, false otherwise
     */
    public boolean canCurrencyDelete(int id) {
        return RealmHelper.canCurrencyDelete(realm, id);
    }

    public void deleteCurrency(int id) {
        realm.beginTransaction();
        realm.where(Currency.class).equalTo(Currency.fId, id).findFirst().deleteFromRealm();
        realm.commitTransaction();
    }

    public void deleteCurrencyNew(int id) {
        if (realm.isInTransaction()) {
            commitTransaction();
        }
        realm.beginTransaction();
        realm.where(Currency.class).equalTo(Currency.fId, id).findFirst().deleteFromRealm();
    }

    public boolean commitTransaction() {
        if (realm.isInTransaction()) {
            realm.commitTransaction();
            return true;
        }
        return false;
    }

    public void cancelTransaction() {
        realm.cancelTransaction();
    }

    /**
     * Save new Book
     *
     * @param title = book title
     * @param desc  = book description (optional)
     */
    public void saveBook(String title, String desc) {
        realm.beginTransaction();
        Book book = realm.createObject(Book.class);
        book.setId(realm);
        book.setTitle(title);
        book.setDesc(desc);
//        book.setTotal(0.0);
        realm.commitTransaction();
    }

    /**
     * get book details by its id
     *
     * @param id = bookId
     * @return = Book
     */
    public Book getBook(int id) {
        return realm.where(Book.class).equalTo(Book.fId, id).findFirst();
    }

    /**
     * Update Book Details
     *
     * @param id    = bookId of Book to be updated
     * @param title = book title
     * @param desc  = book description (optional)
     */
    public void updateBook(int id, String title, String desc) {
        realm.beginTransaction();
        Book book = realm.where(Book.class).equalTo(Book.fId, id).findFirst();
        book.setTitle(title);
        book.setDesc(desc);
        realm.commitTransaction();
    }
}

File: app/src/main/java/com/microlabs/trallet/repo/DatabaseBookRepository.kt
package com.microlabs.trallet.repo

/**
 * Created by Hansen on 4/2/2017.
 *
 *
 * Database of Book and RExpense Repository
 */

open class DatabaseBookRepository {

////    private val realm: Realm = Realm.getDefaultInstance()
//
//    /**
//     * Get All Book List sorted by its Id
//     *
//     * @return list of Book, sort by its Id ascending
//     */
////    open val bookList: RealmResults<Book>? = null
////        get() = realm.where(Book::class.java).sort(Book.fId).findAllAsync()
//
//    /**
//     * Check whether default RCategory is exist
//     * Default RCategory is 4 (Food, Transport, Shop, and Others)
//     *
//     * @return true if default RCategory is NOT exists
//     */
//    open val isDefaultCategoryNotExist: Boolean
//        get() = realm.where(RCategory::class.java).count() < 4
//
//    /**
//     * Check whether default RCurrency is exist
//     * Default RCurrency is 2 (USD $ and IDR Rp)
//     *
//     * @return true if Default RCurrency is NOT exists
//     */
//    open val isDefaultCurrencyNotExist: Boolean
//        get() = realm.where(RCurrency::class.java).count() < 2
//
//    /**
//     * Get All RCategory
//     *
//     * @return = List of RCategory
//     */
//    open val RCategoryList: List<RCategory>
//        get() = realm.where(RCategory::class.java).findAll()
//
//    /**
//     * Get All RCurrency
//     *
//     * @return = List of RCurrency
//     */
//    open val RCurrencyList: List<RCurrency>
//        get() = realm.where(RCurrency::class.java).findAll()
//
//    /**
//     * Insert Default RCategory to Repository if not exists
//     */
//    open fun loadDefaultCategory() {
//        for (i in 0..3) {
//            realm.beginTransaction()
//            val c = realm.createObject(RCategory::class.java)
//            c.setId(realm)
//            when (i) {
//                0 -> c.name = "Foods"
//                1 -> c.name = "Transport"
//                2 -> c.name = "Shop"
//                3 -> c.name = "Others"
//            }
//            realm.commitTransaction()
//        }
//    }
//
//    /**
//     * Insert Default RCurrency if not exists
//     */
//    open fun loadDefaultCurrency() {
//        for (i in 0..1) {
//            realm.beginTransaction()
//            val c = realm.createObject(RCurrency::class.java)
//            c.setId(realm)
//            when (i) {
//                0 -> {
//                    c.name = "IDR Rp"
//                    c.value = 1.0
//                }
//                1 -> {
//                    c.name = "USD $"
//                    c.value = 13500.0
//                }
//            }
//            realm.commitTransaction()
//        }
//    }
//
//    /**
//     * Get List of Expenses from a book
//     *
//     * @param bookId = bookId where List of Expenses is
//     * @return = List of Expenses, sorted by its Id descending (latest is first)
//     */
//    open fun getExpenseList(bookId: Int): List<RExpense> {
//        return realm.where(RExpense::class.java).equalTo(RExpense.fBook, bookId).sort(RExpense.fId, Sort.DESCENDING).findAll()
//    }
//
//    /**
//     * Delete a Book by its Id
//     *
//     * @param bookId = bookId of Book to remove
//     */
//    open fun deleteBook(bookId: Int) {
//        deleteBook(realm, bookId)
//    }
//
//    /**
//     * Delete an RExpense by its Id
//     *
//     * @param bookId    = bookId of RExpense to remove
//     * Used to decrease book total value
//     * @param expenseId = expenseId of RExpense to remove
//     */
//    open fun deleteExpense(bookId: Int, expenseId: Int) {
//        //        RealmHelperKt.setBookDelete(realm, bookId, expenseId);
//        deleteExpense(realm, expenseId)
//    }
//
//    /**
//     * Get RExpense Object by its Id
//     *
//     * @param id = expenseId to get
//     * @return = expense object found
//     */
//    open fun getExpense(id: Int): RExpense {
//        return realm.where(RExpense::class.java).equalTo(RExpense.fId, id).findFirst()!!
//    }
//
//    /**
//     * Save RExpense to database
//     * and set its Book total value
//     *
//     * @param title      = RExpense title
//     * @param bookId     = RExpense bookId
//     * @param value      = RExpense value
//     * @param categoryId = RExpense CategoryId
//     * @param currencyId = RExpense CurrencyId
//     * @param date       = RExpense Date
//     * @param details    = RExpense Details
//     */
//    open fun saveExpense(title: String, bookId: Int, value: Double, categoryId: Int, currencyId: Int, date: Date, details: String) {
//        realm.beginTransaction()
//        val expense = realm.createObject(RExpense::class.java)
//        expense.setId(realm)
//        expense.title = title
//        expense.value = value
//        expense.categoryId = categoryId
//        expense.currencyId = currencyId
//        expense.details = details
//        expense.bookId = bookId
//        expense.date = date
//        realm.commitTransaction()
//        //        RealmHelperKt.setBookTotal(realm, bookId, value);
//    }
//
//    /**
//     * Close Realm Instance to prevent memory leaks
//     */
//    open fun close() {
//        realm.close()
//    }
//
//    /**
//     * Update RExpense Data
//     *
//     * @param expenseId  = expenseId to be updated
//     * @param title      = updated expense title
//     * @param bookId     = expense bookId to update the book total value
//     * @param value      = updated expense value
//     * @param categoryId = updated expense categoryId
//     * @param currencyId = updated expense currencyId
//     * @param date       = updated expense date
//     * @param details    = updated expense details
//     * @param oldValue   = expense old value to update the book total value
//     */
//    open fun updateExpense(expenseId: Int, title: String, bookId: Int, value: Double, categoryId: Int, currencyId: Int, date: Date, details: String, oldValue: Double) {
//        realm.beginTransaction()
//        val expense = realm.where(RExpense::class.java).equalTo(RExpense.fId, expenseId).findFirst()
//        expense?.title = title
//        expense?.value = value
//        expense?.details = details
//        expense?.categoryId = categoryId
//        expense?.currencyId = currencyId
//        expense?.date = date
//        realm.commitTransaction()
//        //        RealmHelperKt.setBookEdit(realm, bookId, oldValue, value);
//    }
//
//    /**
//     * Save New RCurrency
//     *
//     * @param title = RCurrency title
//     * @param value = RCurrency value
//     * @return = true if success to save, false if currency title already exists
//     */
//    open fun saveCurrency(title: String, value: Double): Boolean {
//        if (realm.where(RCurrency::class.java).equalTo(RCurrency.fName, title).findFirst() == null) {
//            realm.beginTransaction()
//            val currency = realm.createObject(RCurrency::class.java)
//            currency.setId(realm)
//            currency.name = title
//            currency.value = value
//            realm.commitTransaction()
//            return true
//        } else {
//            return false
//        }
//    }
//
//    /**
//     * Update RCurrency Details
//     *
//     * @param id    = currencyId of RCurrency to be updated
//     * @param title = currency title
//     * @param value = currency value
//     * @return true if success to update currency, fail if other currency with same title already exists
//     */
//    open fun updateCurrency(id: Int, title: String, value: Double): Boolean {
//        if (realm.where(RCurrency::class.java).equalTo(RCurrency.fName, title).notEqualTo(RCurrency.fId, id).findFirst() != null) {
//            realm.beginTransaction()
//            val currency = realm.where(RCurrency::class.java).equalTo(RCurrency.fId, id).findFirst()
//            currency?.name = title
//            currency?.value = value
//            realm.commitTransaction()
//            return true
//        } else {
//            return false
//        }
//    }
//
//    /**
//     * Check if currency can be removed
//     * currency can only be removed when no longer being used
//     *
//     * @param id = currencyId to be checked
//     * @return true if currency can be removed, false otherwise
//     */
//    open fun canCurrencyDelete(id: Int): Boolean {
//        return canCurrencyDelete(realm, id)
//    }
//
//    open fun deleteCurrency(id: Int) {
////        realm.beginTransaction()
////        realm.where(RCurrency::class.java).equalTo(RCurrency.fId, id).findFirst()?.deleteFromRealm()
////        realm.commitTransaction()
//    }
//
//    open fun deleteCurrencyNew(id: Int) {
//        if (realm.isInTransaction) {
//            commitTransaction()
//        }
//        realm.beginTransaction()
//        realm.where(RCurrency::class.java).equalTo(RCurrency.fId, id).findFirst()?.deleteFromRealm()
//    }
//
//    open fun commitTransaction(): Boolean {
//        if (realm.isInTransaction) {
//            realm.commitTransaction()
//            return true
//        }
//        return false
//    }
//
//    open fun cancelTransaction() {
//        realm.cancelTransaction()
//    }
//
//    /**
//     * Save new Book
//     *
//     * @param title = book title
//     * @param desc  = book description (optional)
//     */
//    open fun saveBook(title: String, desc: String) {
//        realm.beginTransaction()
////        val book = realm.createObject(Book::class.java)
////        book.setId(realm)
////        book.title = title
////        book.desc = desc
//        //        book.setTotal(0.0);
//        realm.commitTransaction()
//    }
//
//    /**
//     * get book details by its id
//     *
//     * @param id = bookId
//     * @return = Book
//     */
//    open fun getBook(id: Int): Book? {
////        return realm.where(Book::class.java).equalTo(Book.fId, id).findFirst()!!
//        return null
//    }
//
//    /**
//     * Update Book Details
//     *
//     * @param id    = bookId of Book to be updated
//     * @param title = book title
//     * @param desc  = book description (optional)
//     */
//    open fun updateBook(id: Int, title: String, desc: String) {
//        realm.beginTransaction()
////        val book = realm.where(Book::class.java).equalTo(Book.fId, id).findFirst()
////        book?.title = title
////        book?.desc = desc
//        realm.commitTransaction()
//    }
}
